### ISSUE TITLE
Strange behavior when trying to update 'json_array' value with QueryBuilder

### ISSUE THREAD

**Issue by ukropich** [doctrine/doctrine2#6426]
I have an entity with a 'json_array' field (for example it's name is 'json_data'). When I try to update this field with an associative array, using the QueryBuilder, like that:
```PHP
$qb
   ->update('SomeEntity', 'e')
   ->set('e.json_data', ':value');
$qb->setParameter(
   'value',
   [
      'array_key' => [
         'some_value1',
         'some_value2'
      ]
   ], 
   'json_array'
);
```
I expect that field will be updated with json:
```JSON
{"array_key":["some_value1","some_value2"]}
```
But I get only the first element of the array: "'some_value1". 
After some search, I've found the method _processParameterValue_ at class _Doctrine\ORM\AbstractQuery_. In this method the following line breaks the array data:
`$value[$key] = is_array($paramValue) ? reset($paramValue) : $paramValue;`

Is it possible to pass a multidimensional array as parameter at QueryBuilder?

**Comment by Ocramius** [comment]
Peculiar: what is being passed down to the DBAL type?

**Comment by ukropich** [comment]
@Ocramius , the string 'json_array' (which is returned by _ClassMetadata->getTypeOfField_ method)

**Comment by Ocramius** [comment]
Yes, and what is being passed to that type (`convertTo*` method calls)?

**Comment by ukropich** [comment]
@Ocramius , i've noticed that 'json_array' is replaced with _Connection::PARAM_STR_ARRAY_, because the method _processParameterValue_ modifies the original value. There is code in the _processParameterMappings_:
```PHP
$value = $this->processParameterValue($value);
$type  = ($parameter->getValue() === $value)
   ? $parameter->getType()
   : ParameterTypeInferer::inferType($value);
```
So _convertToDatabaseValue_ method of the _JsonArrayType_, won't be called.

**Comment by Ocramius** [comment]
Type should probably not be inferred if given...

**Comment by ukropich** [comment]
@Ocramius , yes. And also original array should not be replaced. As I say _processParameterValue_ processing array recursively and taking the first element of inner array values (\Doctrine\ORM\AbstractQuery line 414):
```PHP
if (is_array($value)) {
   foreach ($value as $key => $paramValue) {
      $paramValue  = $this->processParameterValue($paramValue);
      $value[$key] = is_array($paramValue) ? reset($paramValue) : $paramValue;
   }
   return $value;
}
```
